### DISCUSSION TITLE
How to enable two way SSl on and connecting with letsencrypt

### DISCUSSION THREAD

**Discussion by EgHubs** [emqx/emqx#8435] [Q&A] [answered]
So iam using letsencrypt to provide the cert,privkey, ca file
if I need to make a two way SSL certification
I tried to change emqx-example.conf with no hope
so I edited it in the dashboard
<img width="710" alt="Screen Shot 2022-07-06 at 9 21 52 PM" src="https://user-images.githubusercontent.com/73994357/177626764-74cff79a-5590-434e-aeb7-a05894eae4f5.png">

and there I can't connect through the MQTT X and I get the following

> 2022-07-06T18:58:04.410118+00:00 [notice] TLS server: In state wait_cert at tls_handshake_1_3.erl:1475 generated SERVER ALERT: Fatal - Certificate required, -
>  certificate_required
> 2022-07-06T18:58:04.410497+00:00 [error] supervisor: 'esockd_connection_sup - <0.3284.0>', errorContext: connection_shutdown, reason: {ssl_error,{tls_alert,{c
> ertificate_required,"TLS server: In state wait_cert at tls_handshake_1_3.erl:1475 generated SERVER ALERT: Fatal - Certificate required\n certificate_required"
> }}}, offender: [{pid,<0.3385.0>},{name,connection},{mfargs,{emqx_connection,start_link,[#{enable_authn => true,limiter => #{connection => default},listener =>
>  {ssl,default},zone => default}]}}] 
with these configs
<img width="679" alt="Screen Shot 2022-07-06 at 8 59 12 PM" src="https://user-images.githubusercontent.com/73994357/177623239-56671d99-75b8-470b-81f3-541901d7a96a.png">

"Note that ca.pem is the cacert file assigned in emqx.conf which is the o/p of the `§cat` of chain.pem and the ISRG_Root_X1.pem"

The thing is where do I get the Client Certificate File and the Client key file when using letsencrypt
as I digged through the internet and found that letsencrypt doesn't provide these file.

So do I have to create them as a self signed with openssl? and how do i do that?

But if I do that are I'm giving up the security encryption features from letsencrypt or is that just a separate two way verification thing that will menatian the security feature by letsencrypt?

Sorry i know that i already have stated a lot of questions
but how to finally use that into esp8266
I have succefuley connected with esp into the SSL at port 8883 using the following codes from this [link](https://github.com/256dpi/arduino-mqtt/blob/master/examples/AdafruitHuzzahESP8266Secure/AdafruitHuzzahESP8266Secure.ino)

Which states the following

```
  WiFiClientSecure net;
  net.setInsecure();
  client.begin("public.cloud.shiftr.io", 8883, net);
```

but if I need to verify the server from the client side i found the following [gist](https://gist.github.com/eLement87/133cddc5bd0472daf5cb35a20bfd811e#file-mqtt_tls_working-ino-L27)

Which states the following

```
// openssl x509 -in  mqttserver.crt -sha1 -noout -fingerprint
const char* fingerprint = "28:7C:0E:AC:E1:B9:00:9B:DB:4B:BA:19:0A:3D:4A:B5:F6:AA:4D:A6";
```

so in my case will just calling the following work?

`
openssl x509 -in  /etc/letsencrypt/live/domainname.com/fullchain.pem -sha1 -noout -fingerprint
`

**Comment by zmstone** [comment, marked as answer]
You have "Fail If No Client Cert" set to true.
In this case, as the ": Fatal - Certificate required\n certificate_required" is telling us, the client did not send its certificate.

From MQTTX, you'll need to provide the client certificates and client private key files.
For testing, you can use the same private key and certificate file as the server side.

As for how to make esp8266 send client certificates, I'm not sure.

Some more NOTEs: 
To distribute your devices (which has esp8266 running), you'll maybe need to create your own self-signed CA
and issue certificates for this CA.
And this CA should be added to the cacertfile for EMQX so it can verify the client certificate.

I have a collection of scripts to do it:
https://github.com/zmstone/tls-scripts
disclaimer: it's good for prototyping, but for production, you should probably use a more production-ready PKI system.

**Reply by EgHubs** [reply]
Thanks for your quick replays
> From MQTTX, you'll need to provide the client certificates and client private key files. For testing, you can use the same private key and certificate file as the server side.

And for production?


> disclaimer: it's good for prototyping, but for production, you should probably use a more production-ready PKI system.

Can you please guide me for that?, any link for any way?

**Reply by zmstone** [reply]
TBH. I think it's a bit too early to even think about having your own [PKI](https://en.wikipedia.org/wiki/Public_key_infrastructure)
If all clients are programmed by you, and you are not integrating with other vendors:
I would recommend to use openssl command line to generate your own CA and use it to issue both server side and client side certificates. To make things easier, no intermediate CAs.

For the browser facing services, your lets-encrypt certificates should work.

**Reply by EgHubs** [reply]
i will definitely look into that, thanks for your great help [❤️](https://emojipedia.org/red-heart/)

**Comment by zmstone** [comment]
this comment in the gist
https://gist.github.com/eLement87/133cddc5bd0472daf5cb35a20bfd811e#file-mqtt_tls_working-ino-L26
hints that it's the server's certificate CA's fingerprint.
however I'm not sure if it should be the root CA or immediate CA.

You can maybe try to get the fingerprint of the root CA and use  it at L28

**Comment by zmstone** [comment]
https://gist.github.com/eLement87/133cddc5bd0472daf5cb35a20bfd811e#file-mqtt_tls_working-ino-L69-L117
this code block hints that the client private key and certificate and CA files are loaded 
seems to be `.der` format, but not `.pem` format though

**Reply by EgHubs** [reply]
oh didn't notice that, haven't tested it anyway because I couldn't get it to work in MQTT X in the first place
also from the comment on that gist I guess no one has accomplished it :)
But, in a security perspective, do I really need that?
or should I just continue using the following normal way mentioned here without validating the cert files -verify_none-

  ```
WiFiClientSecure net;
  net.setInsecure();
  client.begin("public.cloud.shiftr.io", 8883, net);
```
but there is also a strange thing in the prev [example](https://github.com/256dpi/arduino-mqtt/blob/master/examples/AdafruitHuzzahESP8266Secure/AdafruitHuzzahESP8266Secure.ino#L28)
which are the following comments

```
 // do not verify tls certificate
  // check the following example for methods to verify the server:
  // https://github.com/esp8266/Arduino/blob/master/libraries/ESP8266WiFi/examples/BearSSL_Validation/BearSSL_Validation.ino
  net.setInsecure();
  client.begin("public.cloud.shiftr.io", 8883, net);
```
if you understand these I would be pleased if you explain to me :)

**Reply by zmstone** [reply]
I don't quite understand TBH.
but `net.setInsecure();` is pretty obvious that the client is NOT validating server certificate.

**Reply by EgHubs** [reply]
> I don't quite understand TBH. but `net.setInsecure();` is pretty obvious that the client is NOT validating server certificate.

All that time I was reading it Secure and not Insecure [😂](https://emojipedia.org/face-with-tears-of-joy/)
anyway I don't want to bother you about any arduino library and stuff like that
ill take care of that later
im just asking if I made a broker in a production state without two way SSl
I mean with the current configs with FailIfNoPeerCert sit to false

**Reply by zmstone** [reply]
For the best security, mutual TLS is recommended.
